go/ql/lib/semmle/go/frameworks/Fasthttp.qll

```diff
@@ -0,0 +1,457 @@
+          "SetCanonical"
+        ] and
+      result = this.getArgument(0)
+    }
+
+    override string getHeaderName() {
+      result = Http::HeaderWrite::Range.super.getHeaderName()
+      or
+      methodName = ["SetContentType", "SetContentTypeBytes", "Success", "SuccessString"] and
+      result = "content-type"
+    }
+
+    override DataFlow::Node getValue() {
+      if methodName = ["SetContentType", "SetContentTypeBytes", "Success", "SuccessString"]
+      then result = this.getArgument(0)
+      else result = this.getArgument(1)
+    }
+
+    override Http::ResponseWriter getResponseWriter() {
+      result.(ResponseWriter).getAHeaderObject() = this
+    }
+  }
+
+  private class ResponseWriter extends Http::ResponseWriter::Range {
+    SsaWithFields v;
+
+    ResponseWriter() {
+      this = v.getBaseVariable().getSourceVariable() and
+      (
+        v.getType().hasQualifiedName(packagePath(), ["Response", "ResponseHeader"]) or
+        v.getType().(PointerType).getBaseType().hasQualifiedName(packagePath(), "RequestCtx")
+      )
+    }
+
+    override DataFlow::Node getANode() { result = v.similar().getAUse().getASuccessor*() }
+
+    /** Gets a header object that corresponds to this HTTP response. */
+    DataFlow::MethodCallNode getAHeaderObject() {
+      result.getTarget().getName() =
+        [
+          "Add", "AddBytesK", "AddBytesKV", "AddBytesV", "Set", "SetBytesK", "SetBytesKV",
+          "SetCanonical", "SetContentType", "SetContentTypeBytes", "Success", "SuccessString"
+        ] and
+      this.getANode() = result.getReceiver()
+    }
+  }
+
+  private predicate responseBodyWriterResult(DataFlow::Node src) {
+    exists(Method responseBodyWriter |
+      responseBodyWriter.hasQualifiedName(packagePath(), "Response", "BodyWriter") and
+      src = responseBodyWriter.getACall().getResult(0)
+    )
+  }
+
+  private module ResponseBodyWriterFlow = DataFlow::SimpleGlobal<responseBodyWriterResult/1>;
+
+  private class ResponseBody extends Http::ResponseBody::Range {
+    DataFlow::MethodCallNode call;
+    string methodName;
+
+    ResponseBody() {
+      exists(Method m |
+        m.hasQualifiedName(packagePath(), "Response", methodName) and
+        call = m.getACall() and
+        this = call.getArgument(0)
+        or
+        m.hasQualifiedName(packagePath(), "RequestCtx", ["Success", "SuccessString"]) and
+        call = m.getACall() and
+        this = call.getArgument(1)
+      ) and
+      methodName =
+        [
+          "AppendBody", "AppendBodyString", "SetBody", "SetBodyRaw", "SetBodyStream",
+          "SetBodyString", "Success", "SuccessString"
+        ]
+      or
+      exists(Method write, DataFlow::CallNode writeCall |
+        write.hasQualifiedName("io", "Writer", "Write") and
+        writeCall = write.getACall() and
+        ResponseBodyWriterFlow::flowsTo(writeCall.getReceiver()) and
+        this = writeCall.getArgument(0)
+      ) and
+      methodName = "BodyWriter"
```

Comment:
I see. `DataFlow::SimpleGlobal` isn't suitable then, as you need the beginning and the end of the path, and I don't think it can give that. I will double check. The other options are `DataFlow::Global`, which seems like too much, or local flow, which stays within a function. I see that the other libraries that do this kind of thing use local flow. But I remember from when I recommended `SimpleGlobal` that I searched GitHub and found lots of examples of the return value of `Response.BodyWriter` being passed to other functions before `Write` was called on it.

I think you should start with local flow (either `getASuccessor*` or `DataFlow::localFlow()` and maybe use MRVA to see if that means you are missing a lot. (I mean run MRVA with a query like `can't find local flow to a call to Write()`.

Btw, looking at [the search results](https://github.com/search?q=lang%3AGo+Response.BodyWriter%28%29&type=code) I see that the writer is often passed as the first argument to `fmt.Fprintf`, so you might want to include that as an alternative destination to `io.Writer.Write()`.

Reply:
An idea I just had: if a function has arguments which are a `Writer w` and some other things, then the odds are pretty good that it's going to pass (some of) the other arguments to `w.Write()`. So instead of `fmt.Fprintf()` you could say "the return value flows to an argument of a call to any function, and the response body is any other syntactic argument of that call." You would have to test out how good this heuristic is, either by looking through the search results I linked earlier and seeing if you think it works most of the time, or by coding it up and running MRVA on it. (And you'd have to keep a case for `w.Write()` as well.)

Reply:
Ah, I have just found out that there is a way to adapt the `SimpleGlobal` approach to give you a source and a sink. I'll suggest it in a new review.